_source/_posts/2022-08-24-state-of-project-loom.md

```diff
@@ -0,0 +1,261 @@
+
+Another common use case is parallel processing or multi-threading. You must ensure thread synchronization when executing a parallel task distributed over multiple threads to avoid data races. The implementation becomes even more fragile and puts a lot more responsibility on the developer to ensure there are no issues like thread leaks and cancellation delays.
+
+Project Loom aims to fix these issues in the current concurrency model by introducing two new features, virtual threads, and structured concurrency.
+
+### Virtual Threads
+
+> Java 19 is scheduled to be released in September 2022, and Virtual threads will be a preview feature. Yayyy!
+
+[Virtual threads](https://openjdk.org/jeps/425) are lightweight threads that are not tied to OS threads but are managed by the JVM. They are suitable for thread-per-request programming styles without having the limitations of OS threads. You can create millions of virtual threads without affecting throughput. This is quite similar to coroutines, like [Goroutines](https://go.dev/tour/concurrency/1), made famous by the Go programming language.
+
+The new virtual threads in java 19 will be pretty easy to use. Compare the below with Golang's Goroutines or Kotlin's coroutines.
+
+**Virtual thread**
+
+```java
+Thread.startVirtualThread(() -> {
+    System.out.println("Hello, Project Loom!");
+});
+```
+
+**Goroutine**
+
+```go
+go func() {
+    println("Hello, Goroutines!")
+}()
+```
+
+**Kotlin coroutine**
+
+```kotlin
+runBlocking {
+    launch {
+        println("Hello, Kotlin coroutines!")
+    }
+}
+```
+
+A fun fact is that before Java 1.1, Java had support for green threads (a.k.a virtual threads). The new implementation of virtual threads is done in the JVM, where it maps multiple virtual threads to one or more OS threads, and the developer can use virtual threads or platform threads as per their needs. A few other important aspects of virtual threads are:
+
+- They are a `Thread` in code, runtime, debugger, and profiler
+- It's a Java entity and not a wrapper around a native thread
+- Creating and blocking them are cheap operations
+- They should not be pooled
+- It uses a work-stealing `ForkJoinPool` scheduler
+- Pluggable schedulers can be used for asynchronous programming
+- Virtual thread will have its own stack memory
+- API is very similar to OS threads and hence easier to adopt/migrate
+
+Let's look at some examples to show how powerful virtual threads are.
+
+#### Total number of threads
+
+First, let's see how many platform threads vs. virtual threads we can create on a machine (My machine is Intel Core i9-11900H, 8 cores, 16 threads, and 64GB RAM).
```

Comment:
Does OS matter? If so, you might want to mention it here.

Reply:
Ya could matter a bit